qa/rpc-tests/feature_zip239.py

```diff
@@ -0,0 +1,221 @@
+
+        # Activate NU5. Block height after this is 210.
+        self.nodes[0].generate(9)
+        self.sync_all()
+
+        # Add v4 transaction to the mempool.
+        node1_taddr = self.nodes[1].getnewaddress()
+        opid = self.nodes[0].z_sendmany(sproutzaddr, [{
+            'address': node1_taddr,
+            'amount': 1,
+        }])
+        v4_txid = uint256_from_str(hex_str_to_bytes(
+            wait_and_assert_operationid_status(self.nodes[0], opid)
+        )[::-1])
+
+        # Add v5 transaction to the mempool.
+        v5_txid = self.nodes[0].sendtoaddress(node1_taddr, 1, "", "", True)
+        v5_tx = self.nodes[0].getrawtransaction(v5_txid, 1)
+        assert_equal(v5_tx['version'], 5)
+        v5_txid = uint256_from_str(hex_str_to_bytes(v5_txid)[::-1])
+        v5_auth_digest = uint256_from_str(hex_str_to_bytes(v5_tx['authdigest'])[::-1])
+
+        # Wait for the mempools to sync.
+        self.sync_all()
+
+        # Set up test nodes.
+        # - test_nodes[0] will only request v4 transactions
+        # - test_nodes[1] will only request v5 transactions
+        # - test_nodes[2] will test invalid v4 request using MSG_WTXID
+        # - test_nodes[3] will test invalid v5 request using MSG_TX
+        test_nodes = []
+        connections = []
+
+        for i in range(4):
+            test_nodes.append(TestNode())
+            connections.append(NodeConn(
+                '127.0.0.1', p2p_port(0), self.nodes[0], test_nodes[i],
+                protocol_version=NU5_PROTO_VERSION))
+            test_nodes[i].add_connection(connections[i])
+
+        NetworkThread().start() # Start up network handling in another thread
+        [x.wait_for_verack() for x in test_nodes]
+
+        #
+        # inv
+        #
+
+        # On a mempool request, nodes should return an inv message containing:
+        # - the v4 tx, with type MSG_TX.
+        # - the v5 tx, with type MSG_WTX.
+        for testnode in test_nodes:
+            self.verify_inv(testnode, [
+                self.cinv_for(v4_txid),
+                self.cinv_for(v5_txid, v5_auth_digest),
+            ])
+
+        #
+        # getdata
+        #
+
+        # We can request a v4 transaction with MSG_TX.
+        self.send_data_message(test_nodes[0], v4_txid)
+        self.verify_last_tx(test_nodes[0], v4_txid)
+
+        # We can request a v5 transaction with MSG_WTX.
+        self.send_data_message(test_nodes[1], v5_txid, v5_auth_digest)
+        self.verify_last_tx(test_nodes[1], v5_txid, v5_auth_digest)
+
+        # Requesting with a different authDigest results in a notfound.
+        self.send_data_message(test_nodes[1], v5_txid, 1)
+        self.verify_last_notfound(test_nodes[1], v5_txid, 1)
+
+        # Requesting a v4 transaction with MSG_WTX causes a disconnect.
+        self.send_data_message(test_nodes[2], v4_txid, (1 << 256) - 1)
+        self.verify_disconnected(test_nodes[2])
+
+        # Requesting a v5 transaction with MSG_TX causes a disconnect.
```

Comment:
ZIP 239 seems to be silent on whether this is an error. Should there be a requirement for this similar to the one for requesting a v4 tx using `MSG_WTX`? (I don't think there is any DoS potential here, but that needs thinking about.)

Reply:
It's not silent by my reading:
> A new inv type `MSG_WTX` (0x00000005) is added, for use in both `inv` messages and `getdata` requests, indicating that the hash being referenced is the `wtxid` (i.e. the 64-byte value `txid || auth_digest`). This inv type MUST be used when announcing v5 transactions.

That to me says it is an error to use `MSG_TX` for v5 transactions (though the ZIP is perhaps imprecise by using "announcing", which could maybe be interpreted as allowing it for `getdata` but not `inv`, which would be strange).

Reply:
Yes you're right. I'll make it more explicit though.